crates/runtime/src/memory.rs

```diff
@@ -374,7 +431,104 @@ impl RuntimeLinearMemory for ExternalMemory {
+///
+/// [thread proposal]:
+///     https://github.com/WebAssembly/threads/blob/master/proposals/threads/Overview.md#webassemblymemoryprototypegrow
+#[derive(Clone)]
+pub struct SharedMemory(Arc<SharedMemoryInner>);
+impl SharedMemory {
+    /// Construct a new [`SharedMemory`].
+    pub fn new(ty: wasmtime_types::Memory, tunables: &Tunables) -> Result<Self> {
+        assert!(ty.shared);
+        let plan = MemoryPlan::for_memory(ty, tunables);
+        let (minimum_bytes, maximum_bytes) = Memory::limit_new(&plan, None)?;
+        let mut mmap_memory = MmapMemory::new(&plan, minimum_bytes, maximum_bytes, None)?;
+        let def = LongTermVMMemoryDefinition(mmap_memory.vmmemory());
+        let memory: RwLock<Box<dyn RuntimeLinearMemory>> = RwLock::new(Box::new(mmap_memory));
+        Ok(Self(Arc::new(SharedMemoryInner { memory, ty, def })))
+    }
+
+    /// Wrap an existing [Memory] with the locking provided by a [SharedMemory].
+    pub fn wrap(mut memory: Box<dyn RuntimeLinearMemory>, ty: wasmtime_types::Memory) -> Self {
+        assert!(
+            memory.as_any_mut().type_id() != std::any::TypeId::of::<SharedMemory>(),
+            "cannot re-wrap a shared memory"
+        );
+        let def = LongTermVMMemoryDefinition(memory.vmmemory());
+        Self(Arc::new(SharedMemoryInner {
+            memory: RwLock::new(memory),
+            ty,
+            def,
+        }))
+    }
+
+    /// Return the memory type for this [`SharedMemory`].
+    pub fn ty(&self) -> wasmtime_types::Memory {
+        self.0.ty
+    }
+
+    /// Convert this shared memory into a [`Memory`].
+    pub fn as_memory(self) -> Memory {
+        Memory(Box::new(self))
+    }
+
+    /// Return a mutable pointer to the shared memory's [VMMemoryDefinition].
+    pub fn vmmemory_ptr_mut(&mut self) -> *mut VMMemoryDefinition {
+        &self.0.def.0 as *const _ as *mut _
+    }
+
+    /// Return a pointer to the shared memory's [VMMemoryDefinition].
+    pub fn vmmemory_ptr(&self) -> *const VMMemoryDefinition {
+        &self.0.def.0 as *const _
+    }
+}
+
+struct SharedMemoryInner {
+    memory: RwLock<Box<dyn RuntimeLinearMemory>>,
+    ty: wasmtime_types::Memory,
+    def: LongTermVMMemoryDefinition,
+}
+
+struct LongTermVMMemoryDefinition(VMMemoryDefinition);
+unsafe impl Send for LongTermVMMemoryDefinition {}
+unsafe impl Sync for LongTermVMMemoryDefinition {}
+
+/// Proxy all calls through the [`RwLock`].
+impl RuntimeLinearMemory for SharedMemory {
+    fn byte_size(&self) -> usize {
+        self.0.memory.read().unwrap().byte_size()
+    }
+
+    fn maximum_byte_size(&self) -> Option<usize> {
+        self.0.memory.read().unwrap().maximum_byte_size()
+    }
+
+    fn grow(&mut self, delta_pages: u64, store: &mut dyn Store) -> Result<Option<usize>, Error> {
+        self.0.memory.write().unwrap().grow(delta_pages, store)
+    }
+
+    fn grow_to(&mut self, size: usize) -> Result<()> {
+        self.0.memory.write().unwrap().grow_to(size)
+    }
```

Comment:
This growth operation, on success, needs to perform an atomic increment of the `LongTermVMMemoryDefinition` current length I think?

Reply:
Done in `grow` instead.